runtime/perfetto/db.go

```diff
@@ -155,47 +143,269 @@ func (d *DB) Close() error {
+	// Keep retrying as long as we are getting the "locked" error.
+	for r := retry.Begin(); r.Continue(ctx); {
+		rows, err := d.db.QueryContext(ctx, query, args...)
+		if isLocked(err) {
+			continue
+		}
+		return rows, err
+	}
+	return nil, ctx.Err()
+}
+
+func (d *DB) execDB(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	// Keep retrying as long as we are getting the "locked" error.
+	for r := retry.Begin(); r.Continue(ctx); {
+		res, err := d.db.ExecContext(ctx, query, args...)
+		if isLocked(err) {
+			continue
+		}
+		return res, err
+	}
+	return nil, ctx.Err()
+}
+
+// Serve runs an HTTP server that serves traces stored in the database.
+// In order to view the traces, open the following URL in a Chrome browser:
+//
+//	https://ui.perfetto.dev/#!/?url=http://<hostname>:9001?trace_id=<trace_id>
+//
+// , where <hostname> is the hostname of the machine running this server
+// (e.g., "127.0.0.1"), and <trace_id> is the trace ID in hexadecimal format.
+//
+// Perfetto UI requires that the server runs on the local port 9001. For that
+// reason, this method will block until port 9001 becomes available.
+func (d *DB) Serve(ctx context.Context) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/status", func(w http.ResponseWriter, _ *http.Request) {
+		w.Write([]byte("OK")) //nolint:errcheck // response write error
+	})
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		// Set access control according to:
+		//   https://perfetto.dev/docs/visualization/deep-linking-to-perfetto-ui.
+		w.Header().Set("Access-Control-Allow-Origin", "https://ui.perfetto.dev")
+		traceID := r.URL.Query().Get("trace_id")
+		if traceID == "" {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		spans, err := d.fetchSpans(r.Context(), traceID)
+		if err != nil || len(spans) == 0 {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		traces, err := encodeSpans(spans)
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		w.Write(traces) //nolint:errcheck // response write error
+	})
+	server := http.Server{Handler: mux}
+	ticker := time.NewTicker(time.Second)
+	var warnOnce sync.Once
+	for {
+		select {
+		case <-ticker.C:
+			lis, err := net.Listen("tcp", "localhost:9001")
+			if err != nil {
+				warnOnce.Do(func() {
+					fmt.Fprintf(os.Stderr, "Perfetto server failed to listen on port 9001: %v\n", err)
+					fmt.Fprintf(os.Stderr, "Perfetto server will retry until port 9001 is available\n")
+				})
+				continue
+			}
+
+			ticker.Stop()
+			err = server.Serve(lis)
+			if !errors.Is(err, http.ErrServerClosed) {
+				fmt.Fprintf(os.Stderr, "Failed to serve perfetto backend: %v\n", err)
+			}
+			return
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// encodeSpans encodes the given spans in a format that can be read by the
+// Perfetto UI.
+func encodeSpans(spans []*protos.Span) ([]byte, error) {
+	var buf bytes.Buffer
+	buf.Write([]byte{'['})
+	append := func(event []byte) error {
```

Comment:
Can you pick a name different than append to avoid confusion with the builtin append.